### PR TITLE
Space - Chelsea

### PR DESCRIPTION
**Pull request by ChelseaC13**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Spotlight was in the work model.
Describe how you approached testing that model method. What edge cases did you come up with? | I tested that it works with no works. So, with no works in the database it returns nil.
What are `session` and `flash`? What is the difference between them? |  Session holds data while the browser is open. This is great for login information. Flash holds data for one cycle which is great for showing temporary errors or successes.
What was one thing that you gained more clarity on through this assignment? | working with the MVC and how info flows. 
What is the Heroku URL of your deployed application? | https://chelseasmedia-app.herokuapp.com/ 

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * It is generally better to create a new file per style scope.
+ *
+ */
+
+/* Custom bootstrap variables must be set or imported *before* bootstrap. */
+@import "bootstrap";
+
+@import "**/*";
+
+.orange-header{
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:20]: You shouldn't have class names refer to things like colors in them.  They should be named based on what the sections mean.  

(An example from your code of something with a good class name is "spotlight".)

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    font-weight: bold;
+    font-size: 34px;
+    color: #ff5721 !important;
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:23]: You really should avoid using `!important` if you can.  

Especially in production code.  It will cause headaches if other users are trying to make styles work with yours.  You should rely on precedence instead.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    @work = Work.all
+  end
+
+  def show
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
+    if @work.nil?
+      redirect_to root_path
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:17]: This is the kind of code that makes sense to put in a [controller filter](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/09-intermediate-rails/controller-filters.md).

**Comment by kaidamasaki** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | No controller filters.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | No controller filters.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No fixtures.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a work) | No validations on votes.  (Votes should require both a user and a work.)
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable |
| Concise |
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Good job!  There were a few missed things but well done overall.  😃
